File: netwinform-app/Models/VoteAuditLog.cs
using System;
using System.Threading.Tasks;
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;


namespace netwinform_app.Models
{
    public class VoteAuditLog
    {
        [JsonProperty("id")]
        public string Id { get; set; }
        [JsonProperty("action")]
        public string Action { get; set; }
        [JsonProperty("vote_id")]
        public string VoteId { get; set; }
        [JsonProperty("performed_by_user_id")]
        public string PerformedByUserId { get; set; }
        [JsonProperty("performed_by_user_ip")]
        public string PerformedByUserIp { get; set; }
        [JsonProperty("user_agent")]
        public string UserAgent { get; set; }
        [JsonProperty("old_values")]
        public JObject OldValues { get; set; }
        [JsonProperty("new_values")]
        public JObject NewValues { get; set; }
        [JsonProperty("poll_id")]
        public string PollId { get; set; }
        [JsonProperty("option_id")]
        public string OptionId {  get; set; }
    }
}

File: laravel-app/app/Models/VoteAuditLog.php
<?php

namespace App\Models;

use App\Enums\VoteActions;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\BelongsTo;

class VoteAuditLog extends Model
{
    const UPDATED_AT = null;
    protected $fillable = [
        'id',
        'action',
        'vote_id',
        'performed_by_user_id',
        'performed_by_user_ip',
        'platform',
        'user_agent',
        'old_values',
        'new_values',
        'poll_id',
        'option_id',
    ];
    protected $casts = [
        'old_values' => 'array',
        'new_values' => 'array',
        'action' => VoteActions::class,
        'created_at' => 'datetime',
    ];

    protected static function boot()
    {
        parent::boot();

        // Prevent updates (make logs immutable at application level too)
        static::updating(function ($model) {    
            throw new \Exception('Audit logs cannot be updated');
        });

        // Prevent deletes (make logs immutable at application level too)
        static::deleting(function ($model) {
            throw new \Exception('Audit logs cannot be deleted');
        });
    }

    public function user(): BelongsTo
    {
        return $this->belongsTo(User::class);
    }

    /**
     * Get the vote that was acted upon
     */
    public function vote(): BelongsTo
    {
        return $this->belongsTo(Vote::class);
    }

    /**
     * Get the poll related to this log
     */
    public function poll(): BelongsTo
    {
        return $this->belongsTo(Poll::class);
    }

    /**
     * Get the poll option related to this log
     */
    public function option(): BelongsTo
    {
        return $this->belongsTo(PollOption::class, 'option_id');
    }
}
